posts/tools/test-frame-introduction.md

```diff
@@ -0,0 +1,174 @@
+
+Great, what about the situation when the test fails?
+You would have to simulate the same environment and situation in which it failed.
+But sometimes, it is not that simple to reproduce the exact behavior.
+Because of this, you create another mechanism for collecting logs, description of resources, and many more,
+that will be executed in case of the test failure.
+
+All of these will be mostly stored in the repository where you are writing the tests, which makes sense
+and it is the best possible solution at that moment.
+
+## More repositories, more redundancy
+
+In case we have just one repository where all the testing is done, everything is fine.
+But once we want to extend the product and add new components that will also need system testing,
+we start a fight with redundancy.
+And, one day, you will find yourself copying the testing-related classes across multiple repositories,
+mixing things and always trying to refactor the code without success and without "the light at the end of the tunnel" that
+there will be an end to this.
+At least, that is the situation in which we ended up.
+
+Yes, you can use the test classes from the main repository as a Maven dependency, in case you are releasing it together
+with other modules to the Maven Central (or anywhere else).
+However, there are situations when you cannot do that - one of the examples is cyclic dependency.
+
+Let's say that you have the main repository, containing the code for an operator, and you have a few other repositories for the components that
+are managed by the operator.
+Each of the components has a different release cadence, so having them in the operator's repository is not the correct approach.
+In this particular case, if you would use the "system test" dependency from the operator's repository inside the component's repository (and the component
+is used inside the operator's repository), then you create a cyclic dependency, where the component waits for the release of the operator
+with new things in the test classes, and the operator waits for the component to be released, so it can be included in the controller's code.
+
+Or, you can have one repository for the test logic, that will be shared across multiple repositories, and thanks to that you will remove
+the cyclic dependency.
+Nevertheless, with a new separate repository, there are different challenges.
+You need to think about how well it will be maintained - do you and your team have time to actively contribute to this repository and release a new
+version of it regularly, without breaking the logic of the test suites in each of the repositories?
+How frequent will be the releases?
+What about the situation when you work on different repositories outside of the current organization and this module cannot be imported?
+
+## Test-Frame as the solution
+
+Because Skodjob consists of engineers working on different projects, but with the same passion for testing the integration between the components,
```

Comment:
```suggestion
Because **Skodjob** consists of engineers working on different projects but with the same passion for testing the integration between the components,
```